src/Microsoft.OData.Core/MultipartMixed/ODataMultipartMixedBatchWriter.cs

```diff
@@ -0,0 +1,517 @@
+
+            // important to do this first since it will set up the change set boundary.
+            this.SetState(BatchWriterState.ChangesetStarted);
+            Debug.Assert(this.changeSetBoundary != null, "this.changeSetBoundary != null");
+
+            // write the boundary string
+            ODataMultipartMixedBatchWriterUtils.WriteStartBoundary(this.rawOutputContext.TextWriter, this.batchBoundary, !this.batchStartBoundaryWritten);
+            this.batchStartBoundaryWritten = true;
+
+            // write the change set headers
+            ODataMultipartMixedBatchWriterUtils.WriteChangeSetPreamble(this.rawOutputContext.TextWriter, this.changeSetBoundary);
+            this.changesetStartBoundaryWritten = false;
+        }
+
+        /// <summary>
+        /// Creates an <see cref="ODataBatchOperationRequestMessage"/> for writing an operation of a batch request - implementation of the actual functionality.
+        /// </summary>
+        /// <param name="method">The Http method to be used for this request operation.</param>
+        /// <param name="uri">The Uri to be used for this request operation.</param>
+        /// <param name="contentId">The Content-ID value to write in ChangeSet head.</param>
+        /// <param name="payloadUriOption">The format of operation Request-URI, which could be AbsoluteUri, AbsoluteResourcePathAndHost, or RelativeResourcePath.</param>
+        /// <returns>The message that can be used to write the request operation.</returns>
+        protected override ODataBatchOperationRequestMessage CreateOperationRequestMessageImplementation(string method, Uri uri, string contentId, BatchPayloadUriOption payloadUriOption)
+        {
+            // write pending message data (headers, response line) for a previously unclosed message/request
+            this.WritePendingMessageData(true);
+
+            // Add a potential Content-ID header to the URL resolver so that it will be available
+            // to subsequent operations.
+            // Note that what we add here is the Content-ID header of the previous operation (if any).
+            // This also means that the Content-ID of the last operation in a changeset will never get
+            // added to the cache which is fine since we cannot reference it anywhere.
+            if (this.CurrentOperationContentId != null)
+            {
+                AddToPayloadUriConverter(this.CurrentOperationContentId);
+            }
+
+            this.InterceptException(() => uri = CreateOperationRequestUriWrapper(uri, this.rawOutputContext.MessageWriterSettings.BaseUri));
+
+            // create the new request operation
+            this.CurrentOperationRequestMessage = BuildOperationRequestMessage(
+                this.rawOutputContext.OutputStream,
+                method,
+                uri);
+
+            if (this.changeSetBoundary != null)
+            {
+                this.RememberContentIdHeader(contentId);

```

Comment:
The logic to add the content id should be in the base class, before calling CreateOperationRequestMessageImplementation; no need to do it in the derived class (so we can get rid of RememberContentIdHeader).

Reply:
Make sense to move to base class. Change to private but keep the method because it combines various validations + setting values, and it is used in multiple places of CreateOperationRequest...().

---
In reply to: [144113165](https://github.com/OData/odata.net/pull/960#discussion_r144113165) [](ancestors = 144113165)